File: python/leetcode/402.py
class Solution(object):
    def removeKdigits(self, num, k):
        res = []
        self.help(num, k, res)
        ans = "".join(res).lstrip("0")
        return "0" if not ans else ans

    def help(self, num, n, res):
        if n == 0:
            res.append(num)
            return
        length = len(num)
        if length <= n: return
        idx = 0
        for i in range(1, n+1):
            if num[i] < num[idx]:
                idx = i
        res.append(num[idx])
        newstr = num[idx+1:]
        self.help(newstr, n - idx, res)

File: cpp/leetcode/402.cpp
class Solution {
public:
    string removeKdigits(string num, int k) {
        int n = num.size();
        string res = "";
        int left = n - k;
        for(auto ch : num){
            while(k && !res.empty() && res.back() > ch){
                res.pop_back();
                k--;
            }
            res.push_back(ch);
        }
        res.resize(left); // remove tail 
        while(!res.empty() && res[0] == '0') res.erase(res.begin());
        return res.empty() ? "0" : res;
    }
};